parsec/api/protocol/handshake.py

```diff
@@ -479,14 +482,16 @@ def process_challenge_req(self, req: bytes) -> bytes:
 
         assert isinstance(challenge, bytes)
 
-        # TO-DO remove the else for the next release
-        if self.backend_api_version >= (2, 5):
-            # TO-DO Need to use "BaseSignedData" ?
+        # Provides compatibility with API version 2.4 and below
+        # TODO: Remove once API v2.x is deprecated
+        if self.backend_api_version.version == 2 and self.backend_api_version.revision < 5:
+            answer = self.user_signkey.sign(challenge)
+
+        # Used in API v2.5+ and API v3.x
```

Comment:
This is the part where we've already broken our "compatibility across revision" promise 😄 

So next time we should have create 3.0 instead of 2.5, and then only test on `self.backend_api_version.version` for the vast majority of compatibility handling code 

we would use `self.backend_api_version.revision` only to know if a new command (or, in an existing command,  a new field adding information without changing behavior if it is ignored/not provided) is available. Right ?

Reply:
> This is the part where we've already broken our "compatibility across revision" promise

On the contrary, our rule is fulfilled: we have backward compatibility going both ways across revisions.

I agree that it would have been simpler in terms of code to bump API v3 when we introduced the new signing mechanism but since we made sure that every thing remained compatible on both the server and client side then it's ok. 

> we would use self.backend_api_version.revision only to know if a new command (or, in an existing command, a new field adding information without changing behavior if it is ignored/not provided) is available. Right ?

I guess most changes would work as you describe, but we might need more advanced compatibility in some rare cases (like this one).

Reply:
> On the contrary, our rule is fulfilled: we have backward compatibility going both ways across revisions.

I disagree:
 - we have Parsec implementations that are able to handle both <2.5 and >=2.5 so we have backward compatibility
 - BUT we don't have "any Parsec client implementation that support api v2.x is compatible with any Parsec server implementation thas support api v2.y"

So we have "compatibility across Parsec versions" but not "compatibility across revisions of of a given parsec api version"